x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -842,17 +848,47 @@ protected LogicalPlan rule(LogicalPlan p) {
         }
     }
 
-    /**
-     * A COUNT in a local relation will always be 1.
-     */
-    private static class ReplaceCountInLocalRelation extends OptimizerRule<Aggregate> {
+    static class ReplaceAggregationsInLocalRelations extends OptimizerRule<UnaryPlan> {
 
         @Override
-        protected LogicalPlan rule(Aggregate a) {
-            boolean hasLocalRelation = a.anyMatch(LocalRelation.class::isInstance);
+        protected LogicalPlan rule(UnaryPlan plan) {
+            if (plan instanceof Aggregate || plan instanceof Filter) {
+                // This optimization is only correct if `plan` has a LocalRelation descendant that is not filtered.
+                // Because PushDownAndCombineFilters might push down filters to become descendants of `plan`, this optimization
+                // must never run before PushDownAndCombineFilters.
+                LocalRelation relation = unfilteredLocalRelation(plan.child());
+                if (relation != null) {
+                    long count = (relation.executable() instanceof EmptyExecutable ? 0L : 1L);
+
+                    return plan.transformExpressionsDown(AggregateFunction.class, aggregateFunction -> {
+                        if (aggregateFunction instanceof Count) {
+                            return new Literal(Source.EMPTY, count, DataTypes.LONG);
+                        } else if (count == 0) {
+                            return new Literal(Source.EMPTY, null, aggregateFunction.dataType());
+                        } else {
+                            // note, most aggregation functions have already been substituted by something like
```

Comment:
I think referring to the rule is enough to point out that no further processing is needed - indicating how the aggs are replaced is misleading (since the other rule can change or be replaced all together).